### PR TITLE
[WIP] git: support clone via dulwich backend

### PR DESCRIPTION
**Pull request by pmrowla**
* [ ] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [ ] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

Related to https://github.com/iterative/dvc/issues/2215

Depends on https://github.com/dulwich/dulwich/pull/851, https://github.com/dulwich/dulwich/pull/852

### CODE REVIEW

File: tests/func/test_external_repo.py [hunk 1]
```diff
@@ -125,56 +125,36 @@ def test_relative_remote(erepo_dir, tmp_dir):
...
-        assert shallow
-
-        with external_repo(url, rev="master") as repo:
-            with repo.open_by_relpath("file") as fd:
-                assert fd.read() == "bar"
-
-        assert mock_clone.call_count == 1
-        _, shallow = CLONES[url]
-        assert not shallow
+def test_shallow_clone(mocker):
```
Comment (pmrowla) [inline tests/func/test_external_repo.py:128]: placeholder for now, may need to set up a docker container w/git server and fixed repo data for testing against different git server protocols?

**Comment by skshetry** [comment]
@pmrowla, what's up with this PR? Can we ping the maintainer again on the upstream?

**Comment by pmrowla** [comment]
> @pmrowla, what's up with this PR? Can we ping the maintainer again on the upstream?

discussion in the upstream PR is finally moving again (this is still WIP)

**Comment by skshetry** [comment]
Closing as this has to be fixed in the upstream. I'll try to create this draft in the upstream after we have a decent amount of tests there.

**Pull request closed without merging**